Add missing columns of group manager list

GabrielVidal1 opened the pull request:
Fixes #23 
Add the 3 missing columns of the manager list.
The 'Can Manage' column have an additional widget sometimes (the progress-level) but I'm not sure if this is really necessary.
The other 2 columns are represented by an open or closed lock

smadbe commented on src/app/modules/group/components/group-manager-list/group-manager-list.component.html:
```diff
@@ -11,37 +11,78 @@
+              <div class="slanted-header-content">
+                Can Manage
+              </div>
+            </div>
+          </th>
+          <th style="width: 5rem;">
+            <div class="slanted-header">
+              <div class="slanted-header-content">
+                Can grant group access
+              </div>
+            </div>
+          </th>
+          <th style="width: 5rem;">
+            <div class="slanted-header">
+              <div class="slanted-header-content">
+                Can watch members
+              </div>
```
The problem is that these labels go out of the screen and so are not visible. Maybe we could do it in the other direction? Could you try or suggest an idea?

GabrielVidal1 replied:
In the design, this is made possible by the last, column which is the 'Edit' column, that adds enough space for all the headers to be visible. For now I just added an empty column at the end (that will later be replaced by the 'edit' column)

smadbe replied:
Ok thx. We will still have to fix the column repartition (80% is taken by the second column), but that's something else.

smadbe commented on src/app/modules/group/components/group-manager-list/group-manager-list.component.html:
```diff
@@ -11,37 +11,78 @@
+          <td style="min-width: 110px;">
+            {{ rowData.can_manage }}
+          </td>
+          <td style="width: 5rem;">
```
Why are they clickable?

GabrielVidal1 replied:
fixed, they are no longer clickable

smadbe reviewed (changes requested):
Try to polish the elements as they would be pushed in prod right now. Don't hesitate to ask if you have questions.

GabrielVidal1 commented:
I rebased this branch on master to have the progress-level component and also the latest linter changes.
Also changed the icons showed (now a check and a cross instead of locks)
The text for the 'Can Manage' is now more user friendly (instead of just the value of the corresponding field from the backend)

smadbe commented:
👍 you can merge

GabrielVidal1 commented on src/app/modules/group/components/group-manager-list/group-manager-list.component.ts:
```diff
@@ -22,15 +22,28 @@ export class GroupManagerListComponent implements OnChanges {
+      case 'memberships_and_group':
+        return 'Memberships and group';
+      default:
+        return '';
```
Just one last thing, I don't exactly know what to return if the `can_manage` value is not one of the one of the desired value.

smadbe replied:
Hum indeed, this is not really correct. You cannot really consider that `can_manage` is a string as what is returned by the service is a union type (you "downcast" it, you lose information). You should probably define an exported type in the service for this union type, use this type here, and you will not have to handle any default case.

GabrielVidal1 commented on src/app/modules/group/http-services/get-group-managers.service.ts:
```diff
@@ -7,7 +7,7 @@ export interface Manager {
   name: string,
 
-  can_manage: string,
+  can_manage: 'none' | 'memberships' | 'memberships_and_group',
```
I now restrict the type of `can_manage` to only possible values instead of just a `string`

GabrielVidal1 commented on src/app/modules/group/components/group-manager-list/group-manager-list.component.html:
```diff
@@ -11,37 +11,80 @@
+            {{ rowData.name }}
+          </td>
+          <td style="min-width: 110px;">
+            {{ getManagerLevel(rowData) }}
```
The value of the `can_manage` field is now "translated" here to be more user-friendly

smadbe replied:
You cannot do it this way, otherwise this function will be called hundreds times. No function calls can be done in the rendering. 
One way to solve that is after `.getGroupManagers(this.group.id)`  to map each managers to a dictionary where you have added a "can_manage_as_text" for instance. And so having `managers: Manager[] = [];` -> `managers: (Manager&{can_manage_as_text:string})[] = [];` for instance.
You can check what I did with functions such as "withCodeAdditions" but here it is just something internal to this class, it can probably be done more easily.

GabrielVidal1 replied:
Ok I fixed it (I didn't know how to do this properly).
